### DISCUSSION TITLE
BriefcaseDb.pullAndMergeChanges Function Issue

### DISCUSSION THREAD

**Discussion by nn15296** [iTwin/itwinjs-core#1992] [Q&A] [answered]
When using the BriefcaseDb.pullAndMergeChanges() function, it seems to throw errors in imodeljs version 19.2. I've included the code for it below:

```
  private async updateLocalBriefcase(briefcaseProps: LocalBriefcaseProps, version: IModelVersion): Promise<void> {
    let ctx = await this.createContext();
    const briefcase = await BriefcaseDb.open(ctx, { fileName: briefcaseProps.fileName, readonly: false });
    ctx.enter();
    let count = 0;
    while (true) {
      try {
        console.log(` [${getTimeNow()}]  Updating briefcase to latest changeset...`);
        ctx = await this.createContext();
        console.log("! Before pullAndMerge Function !")
        await briefcase.pullAndMergeChanges(ctx, version);
        console.log("! After pullAndMerge Function !")
        ctx.enter();
        briefcase.close();
        break;
      } catch (err) {
        if (++count === this.errorHandlingMaxTries) {
          throw err;
        } else {
          await this.retryAfterError(err, "Error updating briefcase.", count);
        }
      }
    }
  }
```

The error logs which occur are as shown in screenshot below. Note that I console.log just before the pullAndMergeChanges function and just after, so is it an issue with that function, or am I using it incorrectly?

![image](https://user-images.githubusercontent.com/33571096/128188497-1b88347b-698d-455d-b09f-b955eb63ffe2.png)

**Comment by kabentley** [comment, marked as answer]
Is this an Electron app? Do you have local changes when you call `updateLocalBriefcase`?

**Reply by nn15296** [reply]
It's an agent application. There shouldn't be any local changes I don't think - all the agent does is query information from the briefcase, so there's no editing of it.

**Reply by kabentley** [reply]
We are still trying to reproduce this problem. We have a theory, but don't have a testcase yet. In the meantime, a can you try replacing your code with this:

```ts
private async updateLocalBriefcase(briefcaseProps: LocalBriefcaseProps, version: IModelVersion): Promise<void> {
  const ctx = await this.createContext();
  const db = SnapshotDb.openForApplyChangesets(briefcaseProps.fileName);

  try {
    const target = await BriefcaseManager.changesetFromVersion(ctx, version, db.iModelId);
    await BriefcaseManager.processChangesets(ctx, db, target);
  } finally {
    db.close();
  }
```
and let us know if that works?

**Reply by nn15296** [reply]
Hi, that does indeed seem to work. Likewise, changing the readonly option to true when using BriefcaseDb.open() also seemed to solve the issue. Many thanks!
